### DISCUSSION TITLE
638% Lottie framework size increase in 4.1.0

### DISCUSSION THREAD

**Discussion by Jon889** [airbnb/lottie-ios#2120] [General]
Between Lottie 4.0.1 and 4.1.0 the github release asset xcframework zip size increased from 8.65MB to 18.8MB. 

the iOS-arm64 and tvOS-arm64 binaries have jumped from 1.8MB to 13.3MB (638% increase)

This seems quite extreme, is there a reason for this?

**Comment by calda** [comment]
We use a [script](https://github.com/airbnb/lottie-ios/blob/master/Rakefile#L49) to build the XCFramework in CI. That script didn't change between 4.0.1 and 4.1.0. 

The XCFramework published by 4.0.1 was built by Xcode 14 / Swift 5.7 and was incompatible with consumers using Xcode 13 / Swift 5.5. In 4.1 we updated our CI job to build the XCFramework using Xcode 13.2.1: https://github.com/airbnb/lottie-ios/pull/1922. I think I remember this causing the file size to increase. 

I'm open to suggestions if you have an idea on how to reduce the size of this XCFramework, as long as we maintain existing backwards compatibility guarantees.

**Reply by Jon889** [reply]
I'm surprised that changing the Xcode version would change the size of the actual binary for iOS to increase so much

I'm not an expert on what makes up the binary itself sadly. I see @noahsmartin from emerge tools has contributed to lottie-ios before and they are experts on things like this, maybe they could help?

![Screenshot 2023-07-27 at 17 21 04](https://github.com/airbnb/lottie-ios/assets/389853/fafd54aa-88eb-45ec-8fd0-c18b7f16ed0c)

**Reply by noahsmartin** [reply]
👋 Happy to help with this. Building with bitcode was removed in Xcode 14, so the Xcode 14 xcframework doesn't contain bitcode. The Xcode 13 version still contains it, that is what causes the size difference. Since bitcode gets stripped out before the App Store, this doesn't matter for the sizes of actual iOS apps. If no-one needs the bitcode version of Lottie (which they shouldn't, because apps don't use bitcode anymore) you could explicitly disable bitcode even with the Xcode 13 build. To confirm this is what’s happening you can use otool:

version 4.1.0:
```
otool -l Lottie.framework/Lottie | grep "__LLVM"
  segname __LLVM
   segname __LLVM
```

With version 4.0.1 the same command has no results

**Reply by calda** [reply]
Thanks for the suggestion @noahsmartin!

I'll try disabling bitcode when building the XCFramework from now on: https://github.com/airbnb/lottie-ios/pull/2121

**Reply by calda** [reply]
After disabling bitcode, the XCFramework is only 8.7mb instead of 18.8mb.

**Reply by jshchnz** [reply]
@calda We'd be happy to get the Lottie repo set up so that it runs these size comparisons as a status check on all pull requests so issues like this can be caught automatically. For an amazing open source project like Lottie, we'd be happy to offer our Size Analysis service free of charge :)

**Reply by calda** [reply]
@jshchnz, I'd be open to enabling that functionality in this repo if somebody is willing to post a PR setting it up :)

**Reply by noahsmartin** [reply]
Here's how I enabled it in my fork: https://github.com/noahsmartin/lottie-ios/pull/3/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R22 but this uses fastlane which I don't think is used here. So it might require some work to convert to manually calling the API, and finding the right point to run after an xcarchive gets built.

**Reply by calda** [reply]
I probably don't mind using fastlane for this specific job as long as it's self-contained and doesn't affect the rest of the repo
